WIP Use early preview of .Net Core 3.0

bergmeister opened the pull request:
## PR Summary

There are only 1 failing test on Windows due to Compatibility section in application manifest not being embedded in pwsh.exe (https://github.com/dotnet/core-setup/issues/4963).

.Net Core 3.0 adds one reference assembly and also a `.nupkg.metadata` file for each shipped module containing the hash, which I ignore for the installer.

Removed dependency on rcedit to embed manifest and update assembly file properties as .Net Core 3.0 SDK supports same capabilities from csproj file.  Due to this change, the fileversion and productversion embedded as rouces in pwsh.exe has changed to their desired format defined in the csproj.  Tests were updated to account for this difference.

.Net Core 3.0 supports writing to reserved device names (like CON which writes to the console as expected; didn't test COM or LPT but presumably they would write to the COM and printer ports), so updated code to only check for `CLOCK$` as a reserved device name.

_Breaking Change_ By default, long paths are returned by the FileSystemProvider whereas before, it would retain the short path.   This change is due to a change in .Net Core 3.0 behavior related to long paths which invalidated existing tests that used `$TestDrive` which is built from `$env:TEMP` which has a short path (if the username is > 8 characters which is the case on AzPipelines for Windows).

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

bergmeister committed:
try upgrade to 2.2preview2 of .net core

bergmeister committed:
fix version number of sdk (needs to include build number)

bergmeister committed:
[feature] set runtimeversion and revert changes made to alpine docker image creation process

bergmeister committed:
[feature] Fix NuGet package version of WebListener.csproj test project (build number does not match sdk)

bergmeister committed:
[feature] Fix WiX file components due to newer versions of DLLS

bergmeister committed:
trigger ci

bergmeister committed:
The dotnet-xunit test runner does not work with .Net Core 2.2 and due to problems, they have stopped supporting it, therefore use dotnet test as an alternative
https://github.com/xunit/xunit/issues/1675
https://xunit.github.io/releases/2.4

bergmeister committed:
I forgot that trx format ix VsTest, not NUnit

bergmeister committed:
[feature]

bergmeister committed:
[feature] Fix parameter name due to rename

bergmeister committed:
[feature] revert to use xml test logger again

bergmeister committed:
[feature] forgot some folder->file reverts

bergmeister committed:
[feature] Update Markdig.Signed NuGet package to 0.15.4 due to new release 2h ago

bergmeister committed:
[feature] retrigger CI due to sporadic failure

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore2.2

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore2.2

# Conflicts:
#	build.psm1
#	test/hosting/hosting.tests.csproj

bergmeister committed:
fix merge conflict error to have no diff again in the upstreamed xunit code

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore2.2

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore2.2
Conflicts auto-resolved by Tortoise-Git (white space problem only)
# Conflicts:
#	assets/files.wxs

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore2.2

bergmeister committed:
Update to .Net Core 2.2-preview3

bergmeister committed:
[feature] udpate test project version and appveyor cache

bergmeister committed:
[feature] Fix MSI installer by updating to the new version of mscordaccore and sos due to the upgrade from preview2 to preview3

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore2.2

bergmeister committed:
Upgrade to .Net Core 3.0.100-preview-009806 (latest master)

bergmeister committed:
fix runtime version

bergmeister committed:
fix installer and ignore nupkg info files

bergmeister committed:
[feature]

bergmeister committed:
update ResGen project to .Net Core 3.9 since 2.0 is deprecated and is not working with using netcore3.0

bergmeister committed:
[feature] upgrade typegen to .net core 3.0 as well and update comments

bergmeister committed:
[feature] fix comment

bergmeister committed:
[feature] use .net core 2.1 for web listener

iSazonov commented:
@bergmeister As discussed previously we'll release 6.2 _before_ release .Net Core 3.0 so 6.2 should be on 2.1 version. I'd expect moving to 3.0 at april-may 2019.
First public preview 3.0 was announced for the end of 2018 and release May 2019. Only after release 3.0 we free to use it in the repo. You could keep the branch locally and update it.

bergmeister commented:
Yes, the point of this is rather to check/monitor compatibility with .Net Core 3 early so that the .Net team has time to fix issues *before* they release.

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0

# Conflicts:
#	build.psm1
#	src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs
#	src/ResGen/ResGen.csproj
#	src/TypeCatalogGen/TypeCatalogGen.csproj
#	tools/packaging/project/package.csproj

bergmeister committed:
update WebListener.csproj to use new netcore3.0 NuGet packages

bergmeister committed:
Merge branch 'master' of https://github.com/bergmeister/PowerShell into netcore3.0

SteveL-MSFT commented:
We should definitely move to .Net Core 3.0 as early as possible after 6.2.  It would make sense to use this PR to stage this and we can have it as one of the first PRs merged for 6.3.

SteveL-MSFT committed:
[feature]
fix tests due to differences with .Net Core 3.0

iSazonov commented:
I think we need to move from 3.0 Preview1 to latest build - there is already many updates.

SteveL-MSFT committed:
[feature]
update to 3.0.100-preview-009817

SteveL-MSFT commented:
@iSazonov moving to 3.0.100-preview-010042

iSazonov commented:
Windows path tests failed.

Commit:
[feature]
update to latest rcedit, some test fixes

SteveL-MSFT commented:
The problem on Windows is that `$env:TEMP` for a long username returns a 8.3 path.  Made a few fixes to the tests using [FileInfo] to resolve the 8.3 path to long path.  Doesn't feel right, but not sure what else we can do.  This only affects systems (like VSTS CI on Windows) where the username is > 8 characters thus having a difference where $TestDrive has a 8.3 path due to $env:TEMP but in all other cases, the long path is returned.

In another test, looks like [Environment]::OSVersion returns 6.2 on Win10 rather than 10.0.  This [issue](https://github.com/dotnet/corefx/issues/29395) seems to indicate they don't plan to fix this so we may need to use a different way to validate.

iSazonov commented:
> Doesn't feel right, but not sure what else we can do. 

Is this Pester issue or PowerShell Core issue?

SteveL-MSFT commented:
> Is this Pester issue or PowerShell Core issue?

The behavior I'm seeing is probably technically correct although it's not clear to me why it wasn't failing already.  For a user that has a long username, on Windows, their $env:TEMP folder (which is what Pester uses for $TestDrive) is something like: "c:\Users\longna~1\AppData\local\temp".  The test does something with the filesystem which returns "c:\Users\longnameuser\AppData\local\temp" and the test compares the two which don't match and thus the test fails.  Pester could have a new `Should` to compare paths and say that the two are functionally equivalent, but I wouldn't say it's a Pester problem or PSCore problem since `$env:TEMP` is literally "c:\Users\longna~1\AppData\local\temp"

iSazonov commented:
What API we use for $env:TEMP? It seems it could be fixed in .Net Core.

PaulHigin commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -599,7 +599,7 @@ foo
                 $out | Should -HaveCount 2
                 $out[0] | Should -BeExactly (Get-Item ~).FullName
-                $out[1] | Should -BeExactly "$testdrive"
+                $out[1] | Should -BeExactly [System.IO.FileInfo]::new($testdrive).FullName
```
This should be put into a test helper module function, with a name that makes it clear it is converting to Windows long file name as needed.

SteveL-MSFT replied:
Need to think through what is the right thing to do here.  I don't like the idea of have this sprinkled throughout our test code.  Another option is to change how we get environment variables and always return the long path version since the 8.3 syntax is only relevant on downlevel Windows

SteveL-MSFT replied:
Might be better to have `Join-Path` always return a long path which would also address the test issue (or most of them).  Still trying to understand what changed in .Net Core 3.0 that's causing this

SteveL-MSFT replied:
I've decided that by default, the FileSystemProvider will return long paths

SteveL-MSFT commented:
Opened https://github.com/dotnet/corefx/issues/34506 to cover the OSVersion regression in corefx for 3.0

SteveL-MSFT commented:
The OSVersion issue is somehow the application manifest we embed in pwsh.exe is missing the compatibility section.  Investigating.

SteveL-MSFT commented:
Figured out the OSVersion problem, the application manifest isn't being embedded correctly into pwsh.exe.  Opened https://github.com/dotnet/core-sdk/issues/267

Commit:
[feature]
remove dependency on rcedit as .Net Core 3.0 SDK supports embedding manifest from csproj

Commit:
[feature]

fix merge conflict

iSazonov commented:
>Should just be [System.Environment]::GetEnvironmentVariable(). Which returns the 8.3 version of the path on 3.0 and 2.1 so that hasn't changed.

@SteveL-MSFT I think it is here
https://github.com/pester/Pester/blob/b7904b0aa6c0a1444c904b4b1ec5bdbf9a425545/Functions/TestDrive.ps1#L47-L54

iSazonov commented:
@SteveL-MSFT I run simple C# code in .Net Core 3.0 (ver 3.0.100.10076 )
```c#
Console.WriteLine("Result: {0}", Environment.GetEnvironmentVariable("TEMP"));
```
 and get
`C:\Users\ADMINI~1\AppData\Local\Temp`

Commit:
[feature]
update NormalizePath() to also return long paths

SteveL-MSFT commented:
@iSazonov `$env:TEMP` is defined with the short path (probably for compatibility reasons), so it will return exactly the value of that env var.  So that is expected.

iSazonov commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -88,16 +88,25 @@ public FileSystemProvider()
+            // Use FileInfo to convert short paths to long paths on Windows
+#if !UNIX
+            path = (new FileInfo(path)).FullName;
+#endif
```
Do we really need this? Why we want short names?
I'd follow _new .Net Core behavior_ -  if they can do the "breaking change" we can do too.
The change can slow down PowerShell.

I think a fix should be in Environment.GetEnvironmentVariable
Also I guess that FileInfo return normalized path https://source.dot.net/#System.IO.FileSystem/System/IO/FileInfo.cs,35

SteveL-MSFT replied:
That line is so that we _don't_ have short paths.  Given a short path, it will resolve to the long path.

JeremyKuhne replied:
`Path.GetFullPath` will avoid allocating a `FileInfo` (which just calls that anyway). The `GetFullPath` short path behavior isn't strictly correct btw. it is hugely expensive to check for the long path versions so we don't bother unless we see a tilde in the path. You can, in theory, have a short name that doesn't have a tilde- although I've never seen a real-world occurrence.

I'd recommend that you yank this method completely and straight up call `Path.GetFullPath`. You're going to get inconsistent results otherwise as we collapse separators in `GetFullPath` (which won't happen on Unix here). In addition the replacement here won't do the right thing if UNCs or device paths (e.g. `\\?\`) come in.

iSazonov replied:
> That line is so that we don't have short paths. Given a short path, it will resolve to the long path.

Currently PowerShell allow to process short paths and _preserve_ them. Question is - have/should we keep the behavior or not? I see we don't (in the case we should fix Intellisense too). I think it is for PowerShell Committee.
Also if we want to accept new .Net Core 3.0 behavior I don't see reasons to make exclusion for Environment.GetEnvironmentVariable() - what is backward compatibility for PowerShell scripts we warry about?

SteveL-MSFT replied:
@JeremyKuhne thanks for the tip on `Path.GetFullPath`, will make that change.  As for the slashes, this is existing code.  UNC detection happens somewhere else.

JeremyKuhne replied:
> UNC detection happens somewhere else.

Curious as to where that is. It would be nice if device syntax was properly handled with UNCs. `\\?\UNC\server\share\` is a UNC while `\\?\C:\foo` isn't. :) I've got the logic for handling those cases in .NET.

SteveL-MSFT replied:
@JeremyKuhne it's a pretty [simple check](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/namespaces/FileSystemProvider.cs#L4715)

JeremyKuhne replied:
Yeah, that would pull in things that aren't UNCs.  The logic we use is here: https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs#L143-L188

The correct way to check would be something like `(path.StartsWith("\\\\", StringComparison.Ordinal) && !IsDevice(path)) || IsDeviceUnc(path)`. Or even more correct:

``` C#
path.Length > 2
    && ((IsDirectorySeparator(path[0]) && IsDirectorySeparator(path[1]) && !IsDevice(path)) || IsDeviceUnc(path))
```

Win32 doesn't care about the slash direction- they'll get normalized to backslash correctly. You can see this in action with `dir "//localhost/c$"`.

The values in handling device paths correctly:

1. Allows access to files you can't access otherwise
2. Allows long path access without manifesting & setting reg keys
3. Skips normalization (in the `\\?\` case), which will improve performance a bit

SteveL-MSFT replied:
@JeremyKuhne thanks for the feedback.  Looking again in the code, it appears we have a different [utility function](https://github.com/PowerShell/PowerShell/blob/d80154430d600a52a1d1e3a9f3ecf032b36185d4/src/System.Management.Automation/engine/Utils.cs#L966) relying on the uri class to determine if unc.  We should update usage of that other suspect api to use this one.  I'll take care of that as a separate PR from this one.

SteveL-MSFT replied:
The UNC check is fixed via https://github.com/PowerShell/PowerShell/pull/8680

iSazonov replied:
Perhaps I skipped something but I still think that we could:
```
#if !UNIX 
            return Path.GetFullPath(path); 
#endif
```

Commit:
[feature] address Jeremy's feedback
remove reserved device names that now work with 3.0
update check of file properties to match values from csproj

Commit:
[feature] fix psversion test and test-connection test

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:
```diff
@@ -22,9 +22,10 @@ Describe "Basic FileSystem Provider Tests" -Tags "CI" {
-                             "LPT0", "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"
+            $reservedNamesTests = @(
+                # other formally reserved names work with .Net Core 3.0
+                @{ deviceName = "CLOCK$" }
```
Should we fix IsReservedDeviceName() method?

iSazonov commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -716,9 +716,14 @@ Describe "Console host api tests" -Tag CI {
-        $pwsh.VersionInfo.FileVersion | Should -BeExactly $PSVersionTable.PSVersion.ToString().Split("-")[0]
-        $pwsh.VersionInfo.ProductVersion.Replace("-dirty","") | Should -BeExactly $PSVersionTable.GitCommitId
-        $pwsh.VersionInfo.ProductName | Should -BeExactly "PowerShell Core 6"
+        $pwsh.VersionInfo.FileVersion | Should -Match $PSVersionTable.PSVersion.ToString().Split("-")[0]
```
@SteveL-MSFT it would be good to add that we fixed to the PR description for documenting this in changelog after moving to .Net Core 3.0.

SteveL-MSFT replied:
I just updated the PR description covering this.

iSazonov commented:
@SteveL-MSFT As for compatibility section in manifest - maybe we could accept the breaking change for OS version string or there is other compatibility problems (that we don't catch in tests)?

Commit:
[feature] fix get-process test

Commit:
[feature] update IsReservedDeviceName() method due to .Net Core 3.0 changes

iSazonov commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -936,14 +936,11 @@ internal static bool IsAdministrator()
-            // See if it's the correct length. If it's shorter than CON, AUX, etc, it can't be a device name.
-            // Likewise, if it's longer than 'CLOCK$', it can't be a device name.
+            // If it's longer than 'CLOCK$', it can't be a device name.
             if (((compareName.Length < 3) || (compareName.Length > 6)) &&
```
Seems the condition could be updated too.

SteveL-MSFT replied:
Yes

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0

bergmeister commented:
@SteveL-MSFT Will we revisit the inclusion of WinForms/WPF after this PR? At the moment those binaries do not make it into the `publish` or `ref` folder. I think this is because they have to be opted into via using `<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">` instead of `<Project Sdk="Microsoft.NET.Sdk">`.  and for Winforms e.g. `<UseWindowsForms>true</UseWindowsForms>` in the `.csproj` (at least this was my impression when looking at the examples that `dotnet new winforms` or `dotnet new wpf` create). I tried applying such a change to `TypeCatalogGen.csproj` locally but that did not seem to fix it...
I upstreamed the branch in the meantime to keep up to date with master and avoid integration issues.
Out of curiosity: Why are you force-pushing the commits?

bergmeister commented on tools/packaging/packaging.psm1:
```diff
@@ -2747,9 +2747,12 @@ function Test-FileWxs
         if ($indexedAssetFiles -inotcontains $file)
         {
+            $name = Split-Path -Path $file -Leaf
+            if ($name -eq '.nupkg.metadata') {
```
Is it maybe better to add `'*.nupkg.metadata'` to the blacklist `$dontCopy` when building in build.psm1 [here](https://github.com/PowerShell/PowerShell/blob/master/build.psm1#L2315)

SteveL-MSFT replied:
I think it might be better in build.psm1 which already has a blacklist

SteveL-MSFT commented:
@bergmeister also, we want winforms and WPF

bergmeister committed:
Enable WinForms and WPF

iSazonov commented:
Porting of WinForms and WPF is only at the beginning of the path and it is hardly stable. I would postpone this to the next milestone to reduce the amount of what we need to catch and fix in a short period of time. Moreover, it is unlikely that all APIs will be ported there at once.

SteveL-MSFT commented:
@bergmeister typegen doesn't need it

@iSazonov we're likely changing our support lifecycle to match .Net Core LTS/non-LTS which makes 6.3 a non-LTS release.  In that case, we want to get as much features out as possible from .Net Core so we can test compatibility and open issues

iSazonov commented:
@SteveL-MSFT My thoughts were about free time slots and free resources. We can not do everything at once. It would be nice if you published a public roadmap and a time line for the next stage so that everyone could work together effectively. I hope this helps your team attract more help from the community.

SteveL-MSFT commented:
@iSazonov yes, my plan is to have a new roadmap blog published this month

SteveL-MSFT commented on test/tools/Modules/SelfSignedCertificate/SelfSignedCertificate.psd1:
```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Robert Holt.
```
Why is this being added to the repo?  It's being pulled in by PSGalleryTestModules.csproj

iSazonov replied:
:-) It was your commit 2 week ago.

bergmeister committed:
Move exclusion of '*.nupkg.metadata' to build.psm1 as agreed by @SteveL-MSFT in PR discussion

bergmeister committed:
[feature] Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0

# Conflicts:
#	build.psm1
#	test/tools/WebListener/WebListener.csproj

joeyaiello commented:
@iSazonov agreed on the roadmap. We've been chewing on lots of stuff internally, I'd like to get some stuff squared away around early Feb, will have more details after that for sure.

Commit:
[feature] fix application manifest

Commit:
[feature] remove selfsignedcert module from repo

SteveL-MSFT committed:
[feature] Merge branch 'master' into netcore3.0

SteveL-MSFT commented:
There's still some problem with long paths.  Following up with .Net team to determine what api changed and if intentional as I can't isolate what api is returning different results with 3.0.  Seems like it's a regression and not intentional as other file apis in 3.0 happily return a short path.  Also need to add a test case to handle this (with most recent changes):

```powershell
PS C:\Users\reallylongname\repos\PowerShell> New-PSDrive -PSProvider FileSystem -Root $env:temp -Name temp

Name           Used (GB)     Free (GB) Provider      Root                                               CurrentLocation
----           ---------     --------- --------      ----                                               ---------------
temp              513.65        414.92 FileSystem    C:\Users\REALLY~1\AppData\Local\Te…

PS C:\Users\reallylongname\repos\PowerShell> cd temp:
PS temp:\..\..\..\..\..\>
```

Commit:
[feature] remove system.windows.forms from blacklist

Commit:
[feature] fix Get-Process test due to fileversion change

Commit:
[feature] revert change to resolve paths to long paths

SteveL-MSFT commented:
After discussing with .Net team, the long filename change is not expected.  Reverting the change to always resolve to long filename in filesystemprovider.  Will need to debug to understand what changed in .Net Core 3.0 and open issue in corefx repo.

iSazonov commented on PowerShell.Common.props:
```diff
@@ -62,7 +62,6 @@
             Tracking Issue https://github.com/dotnet/sdk/issues/1557
       -->
       <PackageVersion>$(PSCoreBuildVersion)</PackageVersion>
-      <ApplicatioManifest>..\..\assets\pwsh.manifest</ApplicatioManifest>
```
Should it be new PR?

iSazonov commented on build.psm1:
```diff
@@ -533,7 +533,7 @@ function Restore-PSPester
         [ValidateNotNullOrEmpty()]
         [string] $Destination = ([IO.Path]::Combine((Split-Path (Get-PSOptions -DefaultToNew).Output), "Modules"))
     )
-    Save-Module -Name Pester -Path $Destination -Repository PSGallery -RequiredVersion "4.4.1"
```
I believe updating Pester version should be always another PR. We could do this after 6.2 Preview4 if we want the new version.

SteveL-MSFT replied:
Some of the tests are only failing in VSTS and not locally on my Windows development machine.  Trying to figure out the differences in environment; one of which is the version of Pester in VSTS vs what I have locally.  Was trying to see if a change in Pester contributed to some tests failing.  Since it looks like the same failures, I'll revert this.  However, I'll likely need to add more test tracing to figure this out since it doesn't fail locally.

Commit:
[feature] fix symlink test where `target` member is an array and we need to use ToString() to compare with original path

Commit:
[feature] fix conversion test so longfilename is used

Commit:
[feature] change longfile resolution to more optimal api

SteveL-MSFT commented on test/powershell/Language/Parser/Conversions.Tests.ps1:
```diff
@@ -126,7 +126,8 @@ namespace TestTypeResolution {
         $location = & $powershell -noprofile -command $command
-        $location | Should -Be $cmdletDllPath
+        # need to resolve short filename to long filename using FileInfo
+        $location | Should -Be ([System.IO.FileInfo]::new($cmdletDllPath)).FullName
```
Forgot about that.  Will use GetFullPath()

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0

bergmeister committed:
[feature] Empty commit after pulling in upstream changes from master to resolve trivial merge conflict.

bergmeister committed:
[feature] Fix first half of 'Get-Process for admin.Should support -FileVersionInfo' Pester test

- Use expected value at LHS, which also fixes test due to the way matching works
- Add extra tests for other properties

This will still not make the test pass since the issue with the git commit id still persists

Commit:
[feature] fix get-process test for fileversion check

SteveL-MSFT committed:
Merge branch 'master' into netcore3.0

Commit:
[feature] change get-process -fileversioninfo validation due to difference in CI

Commit:
[feature] reverse check of productversion to gitcommitid

SteveL-MSFT commented:
Looks like only one failure which should be fixed by https://github.com/PowerShell/PowerShell/pull/8761

Commit:
[feature] disable overflow test as it takes too long; use optimal method to get longfile path

bergmeister committed:
[feature] Empty commit to retrigger CI

Commit:
[feature] update files.wxs

Commit:
[feature] move to .Net Core 3.0 preview 2

Commit:
[feature] fix case where typegen has an empty string in the file

iSazonov commented on PowerShell.Common.props:
```diff
@@ -96,8 +96,8 @@
-    <RuntimeFrameworkVersion>3.0.0-preview-27122-01</RuntimeFrameworkVersion>
-    <LangVersion>Latest</LangVersion>
+    <RuntimeFrameworkVersion>3.0.0-preview-27324-05</RuntimeFrameworkVersion>
+    <LangVersion>8.0</LangVersion>
```
Why do we set explicit value If latest should be 8.o now?

SteveL-MSFT replied:
The build failed complaining some language feature was only available in 8.0, making this explicit fixed it.

iSazonov replied:
I wonder - we still do not use 8.0 features.

SteveL-MSFT replied:
@iSazonov true, did a search around the netcore repos and didn't find anything related to this.  This is the error if you leave it as `latest`:

```output
/Users/steve/repos/PowerShell/src/System.Management.Automation/engine/COM/ComUtil.cs(318,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
/Users/steve/repos/PowerShell/src/System.Management.Automation/security/Authenticode.cs(193,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
/Users/steve/repos/PowerShell/src/System.Management.Automation/security/Authenticode.cs(516,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
/Users/steve/repos/PowerShell/src/System.Management.Automation/security/Authenticode.cs(541,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
/Users/steve/repos/PowerShell/src/System.Management.Automation/security/Authenticode.cs(611,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
/Users/steve/repos/PowerShell/src/System.Management.Automation/security/nativeMethods.cs(1070,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
/Users/steve/repos/PowerShell/src/System.Management.Automation/security/nativeMethods.cs(1205,17): error CS1658: Feature 'warning action enable or safeonly' is not available in C# 7.3. Please use language version 8.0 or greater.. See also error CS8370. [/Users/steve/repos/PowerShell/src/System.Management.Automation/System.Management.Automation.csproj]
Execution of { dotnet $Arguments } by build.psm1: line 374 failed with exit code 1
At /Users/steve/repos/PowerShell/build.psm1:1999 char:17
+                 throw $errorMessage
+                 ~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : OperationStopped: (Execution of { dotn\u2026ed with exit code 1:String) [], RuntimeException
+ FullyQualifiedErrorId : Execution of { dotnet $Arguments } by build.psm1: line 374 failed with exit code 1

```

Looks like it's due to use of `#pragma warning enable xxxx`.  Best I could find `latest` points to last released (non-Preview) which would be 7.3.  Seems like a bug in the compiler, opened https://github.com/dotnet/roslyn/issues/32958.  We can keep this workaround for now to make progress.

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0

# Conflicts:
#	assets/files.wxs
#	build.psm1

bergmeister committed:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0
Mainly resolved by keeping the version of the branch due to the runtime update
# Conflicts:
#	PowerShell.Common.props
#	assets/files.wxs
#	global.json
#	test/Test.Common.props
#	test/tools/WebListener/WebListener.csproj

SteveL-MSFT commented:
Reverted attempt to move to newer 3.0 release since it caused more problems

bergmeister commented:
We should try the new preview3 that got published yesterday

iSazonov commented:
Since 6.2-GA was forked we can actively work on the PR.
I think we need rebase and then move to 3.0-Preview3.

bergmeister commented:
I tried to update to `preview3` locally but cannot even do a `dotnet restore`. It seems there are no matching Asp.Net Core Nuget packages. The version available on NuGet [here](https://www.nuget.org/packages/Microsoft.AspNetCore.App/3.0.0-preview3-19153-02) is of version `3.0.0-preview3-19153-02` and does not match the requirements of projects depending on it. I get lots of errors similar to this:
```
C:\PowerShell\src\Microsoft.PowerShell.SDK\Microsoft.PowerShell.SDK.csproj : error NU1603: Microsoft.PowerShell.SDK depends on runtime.win-x64.Microsoft.AspNetCore.App (>= 3.0.0-preview3-27503-5) but runtime.win-x64.Microsoft.AspNetCore.App 3.0.0-preview3-27503-5 was not found. An approximate best match of runtime.win-x64.Microsoft.AspNetCore.App 3.0.0-preview4-19121-08 was resolved. [C:\PowerShell\src\powershell-win-core\powershell-win-core.csproj]
....
C:\PowerShell\src\powershell-win-core\powershell-win-core.csproj : error NU1603: pwsh depends on runtime.win-x64.Microsoft.AspNetCore.App (>= 3.0.0-preview3-27503-5) but runtime.win-x64.Microsoft.AspNetCore.App 3.0.0-preview3-27503-5 was not found. An approximate best match of runtime.win-x64.Microsoft.AspNetCore.App 3.0.0-preview4-19121-08 was resolved.
```

iSazonov commented:
I wonder how we depend on Microsoft.AspNetCore.App?

Commit:
Merge branch 'master' of https://github.com/PowerShell/PowerShell into netcore3.0

# Conflicts:
#	assets/files.wxs

SteveL-MSFT commented:
@PoshChan Please get failures

PoshChan commented:
@SteveL-MSFT, you are not authorized to request test failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

SteveL-MSFT commented:
@PoshChan Please get failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

SteveL-MSFT commented:
@PoshChan Please get failures

PoshChan commented:
@SteveL-MSFT, you are not authorized to request test failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

PoshChan-Staging commented:
@SteveL-MSFT, you are not authorized to request test failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

PoshChan-Staging commented:
@SteveL-MSFT, you are not authorized to request test failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

PoshChan-Staging commented:
@SteveL-MSFT, you are not authorized to request test failures

SteveL-MSFT commented:
@PoshChan Please get failures

SteveL-MSFT commented:
@PoshChan-Staging Please get failures

fmichaleczek commented:
@bergmeister I try to compile your .net v3 branch and found the same symptom.

If I remove 'RuntimeFrameworkVersion' on PowerShell.Common.props and Test.Common.props PowerShell.Common.props, it lets me compile.

I also rollback your change to Microsoft.WindowsDesktop.App

```
dotnet.exe -info 
.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.0.0-preview3-19153-02 [c:\program files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.0.0-preview3-27503-5 [c:\program files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.0.0-preview3-27504-2 [c:\program files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

There is a version's conflict between  all Microsoft.*.App runtimes and PowerShell.Common.props
I read all previous what's news about dotnet and aspnet, i don't found anything interesting.

bergmeister commented:
@fmichaleczek At the moment we are waiting for preview4 to fix problems like that. If you want to try out a working version of netcore3 you could try using a daily build or checkout one of the older commits on this branch that used preview1 or preview2

SteveL-MSFT commented:
Looks like Preview.4 just became available, I'll move the build to that version

SteveL-MSFT committed:
Merge branch 'netcore30' into netcore3.0

SteveL-MSFT commented:
Moving this to a working branch on PS/PS
